### PR TITLE
Maajid/implement hetero

### PR DESCRIPTION
**Pull request by smkarlap**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by MaajidKhan** [comment]
> **Description**: Describe your changes.
> 
> **Motivation and Context**
> 
> * Why is this change required? What problem does it solve?
> * If it fixes an open issue, please link to the issue here.



> **Description**: Describe your changes.
> 
> **Motivation and Context**
> 
> * Why is this change required? What problem does it solve?
> * If it fixes an open issue, please link to the issue here.

**Description:**
The hetero plugin for OpenVINO EP enables computing for inference on one network on several devices. 

**Motivation:**
Purposes to execute networks in heterogeneous mode:
-> To utilize accelerators power and calculate heaviest parts of network on accelerator and 
execute not supported layers on fallback devices like CPU
-> To utilize all available hardware more efficiently during one inference

**context:**
During loading of the network to heterogeneous plugin, network is divided to separate parts 
and loaded to dedicated plugins. Intermediate blobs between these sub graphs are allocated 
automatically in the most efficient way.

**Usecases:**
In a model, for layer's which are not fully supported on OpenVINO are run using the default CPU_Execution_Provider.
Using this feature, we can run the whole graph run on OpenVINO if the layers in the model are supported in any of the
fallback devices given in the device_type of Hetero Build. This may help reduce the inference time overall and as well as
avoid sub graph partitioning, hence we gain time.

File: onnxruntime/core/providers/openvino/openvino_execution_provider.h [hunk 1]
```diff
@@ -44,6 +65,19 @@ struct OpenVINOExecutionProviderInfo {
       #elif defined OPENVINO_CONFIG_VAD_F
       device_type_ = "HETERO:FPGA,CPU";
       precision_ = "FP32";
+      #elif defined OPENVINO_CONFIG_HETERO
+        #ifdef DEVICE_NAME
+          #define DEVICE DEVICE_NAME
+        #endif
+      dev_type = DEVICE;
+      if (dev_type.find("HETERO") == 0) {
```
Comment (smkarlap) [inline onnxruntime/core/providers/openvino/openvino_execution_provider.h:73]: Indent the statements within this if block

File: onnxruntime/core/providers/openvino/openvino_execution_provider.h [hunk 2]
```diff
@@ -13,6 +13,27 @@
...
+    std::stringstream ss(s);
+    std::string item;
+
+    while (getline(ss, item, delim)) {
+        result.push_back(item);
+    }
+    return result;
+}
+
+static std::vector<std::string> parseDevices(const std::string& device_string) {
```
Comment (smkarlap) [inline onnxruntime/core/providers/openvino/openvino_execution_provider.h:27]: parseDevices should also detect invalid device names listed and accordingly return the error (maybe as an empty list). Though this is performed in the build.py, device_type may also be passed in dynamically in runtime, which won't go through the python check. We would need a C++ check here too like you have implemented in [build.py line 257](https://github.com/intel/onnxruntime/pull/47/files#diff-a545fb992b8a8781df05a2c40541f429bfff7e0d8ecc9ef0f5d559994e187814R257).

File: onnxruntime/core/providers/openvino/backends/basic_backend.cc [hunk 3]
```diff
@@ -178,6 +183,12 @@ void BasicBackend::Infer(Ort::CustomOpApi& ort, OrtKernelContext* context) {
   }
   // Get Output tensors
   LOGS_DEFAULT(INFO) << log_tag << "Inference successful";
+#ifndef NDEBUG
+  if (openvino_ep::backend_utils::IsDebugEnabled()) {
+  std::string& hw_target = (global_context_.device_id != "") ? global_context_.device_id : global_context_.device_type;
+  printPerformanceCounts(*infer_request_, std::cout, hw_target, true);
```
Comment (smkarlap) [inline onnxruntime/core/providers/openvino/backends/basic_backend.cc:189]: why do we want to print perf counter info for all debug builds?

**Comment by smkarlap** [comment]
Were you able to test batched inference on VAD-M where the entire model is supported with dynamic batch dimension is supported?

File: onnxruntime/core/providers/openvino/backends/basic_backend.cc (hunk 3 above)
Comment (MaajidKhan) [reply onnxruntime/core/providers/openvino/backends/basic_backend.cc:189]: > why do we want to print perf counter info for all debug builds?

The idea of bringing in this functionality was that the user will actually get to see what layer is run on what device for an Hetero Build. so, I incorporated this for debug mode. Any suggestions @smkarlap ?
Do we want to completely remove this?

**Comment by MaajidKhan** [comment]
> Were you able to test batched inference on VAD-M where the entire model is supported with dynamic batch dimension is supported?

Yes @smkarlap . Verified with Latest changes as well.
Here is where you can track, how I did that:
https://jira.devtools.intel.com/browse/HAFP-1174

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -1007,6 +1007,12 @@ if(onnxruntime_USE_OPENVINO)
     add_definitions(-DOPENVINO_CONFIG_VAD_F=1)
   endif()
 
+  if(onnxruntime_USE_OPENVINO_HETERO)
+    add_definitions(-DOPENVINO_CONFIG_HETERO=1)
+    add_definitions(-DDEVICE_NAME="${onnxruntime_USE_OPENVINO_DEVICE}")
+    message("${DEVICE_NAME}")
```
[the remark this answers is not in the document]
Comment (MaajidKhan) [reply cmake/CMakeLists.txt]: > Please Remove message after your are done with validation

Done.

File: onnxruntime/core/providers/openvino/openvino_execution_provider.h (hunk 1 above)
Comment (MaajidKhan) [reply onnxruntime/core/providers/openvino/openvino_execution_provider.h:92]: > Indent the statements within this if block

Done

File: onnxruntime/core/providers/openvino/openvino_execution_provider.h (hunk 2 above)
Comment (MaajidKhan) [reply onnxruntime/core/providers/openvino/openvino_execution_provider.h:36]: @smkarlap  Implemented, changes pushed.

File: onnxruntime/core/providers/openvino/backends/backend_factory.cc [hunk 5]
```diff
@@ -17,10 +17,10 @@ BackendFactory::MakeBackend(const ONNX_NAMESPACE::ModelProto& model_proto,
                             GlobalContext& global_context,
                             const SubGraphContext& subgraph_context) {
   std::string type = global_context.device_type;
-  if (type == "CPU" || type == "GPU" || type == "MYRIAD" || type == "HETERO:FPGA,CPU") {
-    return std::make_shared<BasicBackend>(model_proto, global_context, subgraph_context);
-  } else if (type == "HDDL") {
+  if (type.find("HDDL") != std::string::npos) {
     return std::make_shared<VADMBackend>(model_proto, global_context, subgraph_context);
+  } else if (type == "CPU" || type == "GPU" || type == "MYRIAD" || type == "HETERO:FPGA,CPU" || type.find("HETERO") != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (MaajidKhan) [reply onnxruntime/core/providers/openvino/backends/backend_factory.cc]: Makes sense. Removed it!!

**Pull request closed without merging**